### PR TITLE
Refactor avifImageRGBToYUVLibYUV8bpc() and avif.h

### PR DESCRIPTION
**Pull request by y-guyon**
Describe RGB-to-YUV before YUV-to-RGB in avif.h
Order description and code by avifRGBFormat, avifPixelFormat, and
avifRange for both RGB-to-YUV and YUV-to-RGB.

Fix libyuv RGB-to-YUV 4:0:0 being used for any avifMatrixCoefficients
and not only for BT.601.

Note: Previous description in avif.h was inaccurate for 10-bit YUV-to-RGB and missing for 12-bit
(10-bit has fewer paths than 8-bit and 12-bit got added in https://github.com/AOMediaCodec/libavif/pull/1059).

### CODE REVIEW

File: include/avif/avif.h [hunk 1]
```diff
@@ -529,31 +529,37 @@ AVIF_API void avifImageStealPlanes(avifImage * dstImage, avifImage * srcImage, a
...
+//   *      x     /x/[1|2|5|6|9]/[Limited|Full]
+//   * [1|2|5|6|9]/x/    12     /[Limited|Full]
+// * One of the following combinations (avifRGBFormat from avifPixelFormat/depth):
+//   * RGB      from  [      |      |YUV420|      ]/8-bit
+//   * RGBA     from  [YUV444|YUV422|YUV420|YUV400]/8-bit or [YUV444|YUV422|YUV420]/10-bit or [YUV420]/12-bit
+//   * ARGB     from  [      |YUV422|YUV420|      ]/8-bit
+//   * BGR      from  [      |      |YUV420|      ]/8-bit
+//   * BGRA     from  [YUV444|YUV422|YUV420|YUV400]/8-bit or [YUV444|YUV422|YUV420]/10-bit or [YUV420]/12-bit
+//   * ABGR     from  [      |YUV422|YUV420|      ]/8-bit
+//   * RGB_565  from  [      |YUV422|YUV420|      ]/8-bit
```
Comment (wantehchang) [inline include/avif/avif.h:562]: I suggest we omit the detailed documentation and just say "if there is a libyuv function for the conversion".

I really appreciate the work you did to synthesize these tables, but the tables are starting to become too complicated, and it's hard to keep them up to date.

File: src/reformat_libyuv.c [hunk 2]
```diff
@@ -135,101 +135,68 @@ avifResult avifImageRGBToYUVLibYUV8bpc(avifImage * image, const avifRGBImage * r
...
-            }
-        } else if (rgb->format == AVIF_RGB_FORMAT_ARGB) {
-            if (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV420) {
-                if (image->yuvRange == AVIF_RANGE_LIMITED) {
-                    RGBtoYUV = BGRAToI420;
+        if ((image->yuvFormat == AVIF_PIXEL_FORMAT_YUV444) || (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV422) ||
+            (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV420)) {
+            // Lookup table for RGB To YUV Matrix (average filter).
+            typedef int (*RGBtoYUV)(const uint8_t *, int, uint8_t *, int, uint8_t *, int, uint8_t *, int, int, int);
+            // Third dimension is for avifRange: 0 is AVIF_RANGE_LIMITED, 1 is AVIF_RANGE_FULL.
```
Comment (wantehchang) [inline src/reformat_libyuv.c:145]: Optional: The "0 is AVIF_RANGE_LIMITED, 1 is AVIF_RANGE_FULL" part could be omitted, because they are the numeric values of `AVIF_RANGE_LIMITED` and `AVIF_RANGE_FULL`.

Also at line 174.

File: src/reformat_libyuv.c [hunk 3]
```diff
@@ -135,101 +135,68 @@ avifResult avifImageRGBToYUVLibYUV8bpc(avifImage * image, const avifRGBImage * r
...
-                 image->yuvPlanes[AVIF_CHAN_U],
-                 image->yuvRowBytes[AVIF_CHAN_U],
-                 image->yuvPlanes[AVIF_CHAN_V],
-                 image->yuvRowBytes[AVIF_CHAN_V],
-                 image->width,
-                 image->height)) {
-        return AVIF_RESULT_REFORMAT_FAILED;
-    }
-    return AVIF_RESULT_OK;
+    return AVIF_RESULT_NOT_IMPLEMENTED;
```
Comment (wantehchang) [inline src/reformat_libyuv.c:199]: The TODO comment at line 198 probably should be moved to a different place. It looks out of place now.

File: src/reformat_libyuv.c [hunk 4]
```diff
@@ -555,6 +498,31 @@ avifResult avifImageYUVToRGBLibYUV8bpc(const avifImage * image,
             }
             return AVIF_RESULT_OK;
         }
+    } else if (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV400) {
```
Comment (wantehchang) [inline src/reformat_libyuv.c:501]: Since we are checking for specific values of `image->yuvFormat` now, we should add an `else` branch and return `AVIF_RESULT_INVALID_ARGUMENT` or `AVIF_RESULT_REFORMAT_FAILED`.

On the other hand, if we reach this function, we already know `image->yuvFormat` is one of the four YUV formats in AV1 (it's been validated by the `avifPrepareReformatState()` call in `avifImageRGBToYUV()`, so the original `if (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV400) { ... } else { ...}` structure is correct and simpler. We don't need to validate `image->yuvFormat` redundantly here.

Comment (y-guyon) [reply src/reformat_libyuv.c:501]: I believe following enum order is simpler to maintain, so I just removed `if (image->yuvFormat == AVIF_PIXEL_FORMAT_YUV400)`.

File: src/reformat_libyuv.c (hunk 3 above)
Comment (y-guyon) [reply src/reformat_libyuv.c:199]: It is right after the `if ((image->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_BT470BG) || (image->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_BT601)) {...}` condition. There could be `else if (image->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_IDENTITY) {...}` here, so it looks correctly placed to me. Where would you suggest to move it?

File: src/reformat_libyuv.c [hunk 5]
```diff
@@ -255,49 +222,37 @@ avifResult avifImageYUVToRGBLibYUV(const avifImage * image, avifRGBImage * rgb)
     // Find the correct libyuv YuvConstants, based on range and CP/MC
     const struct YuvConstants * matrixYUV = NULL;
     const struct YuvConstants * matrixYVU = NULL;
-    if (image->yuvRange == AVIF_RANGE_FULL) {
+    if (image->yuvRange == AVIF_RANGE_LIMITED) {
```
Comment (wantehchang) [inline src/reformat_libyuv.c:225]: If I understand it correctly, all the remaining changes are just to test `image->yuvRange` or `image->yuvFormat` in the order of the enum values. I think it's not necessary to make this kind of change.

Also, to separate monochrome from non-monochrome YUV formats, it is cheaper to test `image->yuvFormat == AVIF_PIXEL_FORMAT_YUV400` (the only monochrome format) than testing three formats.

File: include/avif/avif.h [hunk 6]
```diff
@@ -528,32 +528,9 @@ AVIF_API void avifImageStealPlanes(avifImage * dstImage, avifImage * srcImage, a
...
-// * rgb.chromaDownsampling: AVIF_CHROMA_DOWNSAMPLING_AUTOMATIC, AVIF_CHROMA_DOWNSAMPLING_FASTEST
-// * One of the following combinations (avifRGBFormat to avifPixelFormat/MC/Range):
-//   *  BGRA            to  YUV400        /  x  /[Full|Limited]
-//   *  BGRA            to [YUV420|YUV422]/[5|6]/[Full|Limited]
-//   *  BGRA            to  YUV444        /[5|6]/ Limited
-//   *  BGR             to  YUV420        /[5|6]/[Full|Limited]
-//   * [RGBA|ARGB|ABGR] to  YUV420        /[5|6]/ Limited
+// If libavif is built with a version of libyuv offering a fast conversion between RGB and YUV for
+// the given settings (avifMatrixCoefficients, avifColorPrimaries, avifPixelFormat, avifRGBFormat,
+// avifRange, bit depth), libavif will use it. See reformat_libyuv.c for the details.
```
Comment (wantehchang) [inline include/avif/avif.h:533]: I like the new comment. Thanks.

Nit: It doesn't seem necessary to enumerate all the settings we depend on. (It is already quite long but still missing chromeUpsampling/chromaDownsampling/conversion flags.)  It should suffice to just say "for the given inputs".

**Review by wantehchang — APPROVED** [review]
LGTM with comments and optional suggested changes. Also please wait for Vignesh's review.

File: src/reformat_libyuv.c [hunk 7]
```diff
@@ -255,49 +221,37 @@ avifResult avifImageYUVToRGBLibYUV(const avifImage * image, avifRGBImage * rgb)
...
```
Comment (vigneshvg) [inline src/reformat_libyuv.c:224]: i think the reason behind the changes in this function is to order the if checks in the enum order?

if so, it could be a separate PR.

File: src/reformat_libyuv.c [hunk 8]
```diff
@@ -135,101 +135,67 @@ avifResult avifImageRGBToYUVLibYUV8bpc(avifImage * image, const avifRGBImage * r
...
+            RGBtoYUV lutRgbToYuv[AVIF_RGB_FORMAT_COUNT][AVIF_PIXEL_FORMAT_COUNT][2] = {
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { NULL, NULL } },                 // RGB
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { ABGRToI420, avifABGRToJ420 }, { NULL, NULL } }, // RGBA
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { BGRAToI420, NULL }, { NULL, NULL } },           // ARGB
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { RGB24ToI420, RGB24ToJ420 }, { NULL, NULL } },   // BGR
+                { { NULL, NULL }, { ARGBToI444, NULL }, { ARGBToI422, ARGBToJ422 }, { ARGBToI420, ARGBToJ420 }, { NULL, NULL } }, // BGRA
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { RGBAToI420, NULL }, { NULL, NULL } }, // ABGR
+                { { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { NULL, NULL }, { NULL, NULL } }        // RGB_565
+            };
+            RGBtoYUV rgbToYuv = lutRgbToYuv[rgb->format][image->yuvFormat][image->yuvRange];
```
Comment (vigneshvg) [inline src/reformat_libyuv.c:179]: since we are now using image->yuvRange in the lookup tables (as opposed to if/switch statements), we would not have to validate it in avifPrepareReformatState similar to how we validate yuvFormat and rgb->format [1].

[1] https://github.com/AOMediaCodec/libavif/blob/main/src/reformat.c#L31

**Review by vigneshvg — CHANGES_REQUESTED** [review]

File: src/reformat_libyuv.c (hunk 8 above)
Comment (vigneshvg) [reply src/reformat_libyuv.c:179]: *we would now

File: src/reformat_libyuv.c (hunk 7 above)
Comment (wantehchang) [reply src/reformat_libyuv.c:224]: In general, handling enum values in the enum order makes it easier to verify all enum values are handled. In this particular case, the order doesn't matter because it's clear that range is either limited or full. This is why I think it's not necessary to reorder this if-else statement.

Comment (vigneshvg) [reply src/reformat_libyuv.c:224]: yeah i am okay with not re-ordering this as well.

**Review by vigneshvg — APPROVED** [review]

**Comment by vigneshvg** [comment]
lgtm. thanks!

**Review by wantehchang — APPROVED** [review]
Still LGTM!
